qute.jdt/com.redhat.qute.jdt/src/main/java/com/redhat/qute/jdt/internal/template/QuteSupportForTemplateGenerateMissingJavaMemberHandler.java

```diff
@@ -0,0 +1,671 @@
+				lineDelimiter);
+		String cuContent = constructCUContent(cu, typeStub, lineDelimiter);
+		TextDocumentEdit tde = new TextDocumentEdit();
+		tde.setTextDocument(new VersionedTextDocumentIdentifier(uri, 0));
+		tde.setEdits(Arrays.asList(new TextEdit(PREPEND_RANGE, cuContent)));
+		return tde;
+	}
+
+	/*
+	 * Copied & modified from JDT-LS
+	 */
+	private static String constructTypeStub(ICompilationUnit parentCU, String name, int modifiers, String methodName,
+			String methodParamFullyQualifiedType, String lineDelimiter) {
+		StringBuilder buf = new StringBuilder();
+
+		buf.append("@");
+		buf.append(QuteJavaConstants.TEMPLATE_EXTENSION_ANNOTATION);
+		buf.append(lineDelimiter);
+
+		buf.append(Flags.toString(modifiers));
+		if (modifiers != 0) {
+			buf.append(' ');
+		}
+
+		buf.append("class ");
+		buf.append(name);
+		buf.append(" {").append(lineDelimiter); //$NON-NLS-1$
+		buf.append(constructMethodStub(parentCU, methodName, methodParamFullyQualifiedType, lineDelimiter));
+		buf.append('}').append(lineDelimiter);
+		return buf.toString();
+	}
+
+	private static String constructMethodStub(ICompilationUnit compilationUnit, String methodName,
+			String methodParamFullyQualifiedType, String lineDelimiter) {
+		StringBuilder buf = new StringBuilder("\tpublic static String ");
+		buf.append(methodName).append("(").append(methodParamFullyQualifiedType);
+		buf.append(" ");
+		buf.append(getParamNameFromFullyQualifiedType(methodParamFullyQualifiedType));
+		buf.append(") {").append(lineDelimiter);
+		buf.append("\t\treturn null;").append(lineDelimiter);
+		buf.append("\t}").append(lineDelimiter);
+		return buf.toString();
+	}
+
+	/**
+	 * Copied from JDT-LS
+	 */
+	private static String constructCUContent(ICompilationUnit cu, String typeContent, String lineDelimiter)
+			throws CoreException {
+		String fileComment = CodeGeneration.getFileComment(cu, lineDelimiter);
+		String typeComment = CodeGeneration.getTypeComment(cu, cu.getElementName(), lineDelimiter);
+		IPackageFragment pack = (IPackageFragment) cu.getParent();
+		String content = CodeGeneration.getCompilationUnitContent(cu, fileComment, typeComment, typeContent,
+				lineDelimiter);
+		if (content != null) {
+			ASTParser parser = ASTParser.newParser(IASTSharedValues.SHARED_AST_LEVEL);
+			parser.setProject(cu.getJavaProject());
+			parser.setSource(content.toCharArray());
+			CompilationUnit unit = (CompilationUnit) parser.createAST(null);
+			if ((pack.isDefaultPackage() || unit.getPackage() != null) && !unit.types().isEmpty()) {
+				return content;
+			}
+		}
+		StringBuilder buf = new StringBuilder();
+		if (!pack.isDefaultPackage()) {
+			buf.append("package ").append(pack.getElementName()).append(';'); //$NON-NLS-1$
+		}
+		buf.append(lineDelimiter).append(lineDelimiter);
+		buf.append(typeContent);
+		return buf.toString();
+	}
+
+	private static CompilationUnit createQuickFixAST(ICompilationUnit compliationUnit) {
```

Comment:
There is ClasCastException error when this method is called with binary like java.lang.String (it is a IClassFile and not a ICompilationUnit).

I think createQuickFixAST should return null if it is a binary (ex : we try to add a new getter, property in java.lang.String) and otherwise the CompilationUnit.

```java
CompilationUnit createQuickFixAST(IType javaType) {
   if (javaType.isBinary()) {
      return null;
   }
   return ASTResolving.createQuickFixAST((ICompilationUnit) javaType, null);
} 
```